### PR TITLE
Miscelaneous updates (second batch)

### PR DESCRIPTION
**Pull request by talassio**
Second (+10 commits, Jan 29 2019 to Feb 18 2019) mainly related to new printing proposal
      * `Write(out, ('...'))...` is replaced by `buildmsg`/`printmsg` pair with 
      * `Write(rec(i), Fmt=NNNNN)...`
      * Only uses `options%out` and not `options%error`. See note.
      * No more Go To labels, except 100 for exit
      * Comply with LAPACK interface description

**Note** these commits don't address possible issue with `options%error` i/o unit output, future commit will address issue.

**Note** this batch breaks STP

### CODE REVIEW

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 1]
```diff
@@ -41,7 +65,13 @@ module ral_nlls_workspaces
 !    summary for every iteration, = 2 gives a summary of the inner iteration
 !    for each iteration, >= 3 gives increasingly verbose (debugging) output
 
-     INTEGER :: print_level = 0
+     INTEGER :: print_level = 3
```
Comment (tyronerees) [inline libRALFit/src/ral_nlls_workspaces.f90]: The default should be no printing (most people will use this as a black-box deep in code, and won't really care what the algorithm is doing)

File: libRALFit/src/ral_nlls_workspaces.f90 [hunk 2]
```diff
@@ -571,14 +584,14 @@ type, public :: get_svd_J_work ! workspace for subroutine get_svd_J
   type, public :: NLLS_workspace ! all workspaces called from the top level
      logical :: allocated = .false.
      integer :: first_call = 1
-     integer :: iter = 0 
+     integer :: iter = 0
      real(wp) :: normF0, normJF0, normF, normJF
      real(wp) :: normJFold, normJF_Newton
      real(wp) :: Delta
-     real(wp) :: normd
+     real(wp) :: normd = 1.0e10_wp
```
Comment (tyronerees) [inline libRALFit/src/ral_nlls_workspaces.f90]: Why is `normd` set to this value here?

File: libRALFit/example/C/nlls_example.c [hunk 3]
```diff
@@ -73,6 +73,7 @@ int main(void) {
    // Initialize options values
    struct ral_nlls_options options;
    ral_nlls_default_options(&options);
+   options.print_level=0;
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/example/C/nlls_example.c]: Ok

File: libRALFit/src/ral_nlls_internal.f90 [hunk 4]
```diff
@@ -775,87 +894,22 @@ recursive subroutine nlls_iterate(n, m, X,                   &
...
-5020 continue
-    if (options%print_level >= 2) then
-       write(options%out,'(a,i0)') 'RAL_NLLS converged (on step length test) at iteration ', &
-            w%iter
-    end if
-    goto 4000
-    
-    
+8000 Format(85('-'))
+9000 Format(1X,' Iter |  error   |    grad    |  rel grad  |  Delta  |   rho   |S2IF| inn it|  step')
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: `1000 FORMAT('iter',4x,'2nd order?',2x,'Delta',9x,'rho',11x,'0.5||f||^2',4x,'||J''f||',7x,'||J''f||/||f||')`

File: libRALFit/example/Fortran/nlls_example.f90 [hunk 5]
```diff
@@ -98,6 +98,7 @@ program nlls_example
    real(wp), allocatable :: x(:)
    type(params_type) :: params
 
+   options%print_level = 0
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/example/Fortran/nlls_example.f90]: Given new scheme original level=1 corresponds to level=4 of the new scheme.

File: libRALFit/example/Fortran/nlls_example2.f90 [hunk 6]
```diff
@@ -137,12 +137,11 @@ program nlls_example2
...
    
    options%model = 4
-   options%type_of_method = 3
+   options%type_of_method = 2
    options%nlls_method = 4
    options%exact_second_derivatives = .true. 
    options%maxit = 50
-
-   options%print_level = 1
+   options%print_level = 4
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/example/Fortran/nlls_example2.f90]: Given new scheme original level=1 corresponds to level=4 of the new scheme.

File: libRALFit/test/example_module.f90 [hunk 7]
```diff
@@ -1655,6 +1654,9 @@ subroutine switch_to_quasi_newton_tests(options,fails)
      options%type_of_method = 2
      options%reg_order = 0.0_wp
      call setup_workspaces(w,n,m,options,inform)
+     ! switch_to_quasi_newton expect for w%hf_temp to be defined.
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/test/example_module.f90]: Comment typo fixed

File: libRALFit/test/nlls_c_test.c [hunk 8]
```diff
@@ -80,6 +80,7 @@ int generic_test(int model, int method){
   
   options.model = model;
   options.nlls_method = method;
+  options.print_level = 0;
```
Comment (talassio) [inline libRALFit/test/nlls_c_test.c]: Removed.

File: libRALFit/src/ral_nlls_workspaces.f90 (hunk 1 above)
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: Changed: print level is by default zero

File: libRALFit/src/ral_nlls_workspaces.f90 (hunk 2 above)
Comment (talassio) [reply libRALFit/src/ral_nlls_workspaces.f90]: `normd` is now undefined and `test_convergence` for iteration 1 only tests for objective and gradient conditions.

File: libRALFit/src/ral_nlls_internal.f90 [hunk 9]
```diff
@@ -122,118 +108,107 @@ RECURSIVE SUBROUTINE NLLS_SOLVE( n, m, X,                   &
...
+       call nlls_strerror(inform)
+     End If
+     If (buildmsg(1,.False.,options)) then
+       Call print_bye(options,inform)
+     End If
 
-5000 FORMAT(/,'RAL_NLLS failed to converge in the allowed number of iterations')
-5001 FORMAT('ERROR:',1X,A)
+6000 Format(1X,53('-'))
+6001 Format(2X,'RALFit, unconstrained nonlinear least square solver')
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: Updated

File: libRALFit/src/ral_nlls_internal.f90 [hunk 10]
```diff
@@ -430,99 +436,150 @@ recursive subroutine nlls_iterate(n, m, X,                   &
...
-       
-       rho  = -one ! intialize rho as a negative value
 
-       if (options%print_level > 0 ) then 
-          write(options%out,1010) w%iter, ' ', w%Delta, rho, inform%obj, &
-            inform%norm_g, inform%scaled_g
-       end if
+       rho  = -1.0_wp ! intialize rho as a negative value
+
+!      @TAG:ITER0BNR
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: Added comment, details are found in tag ITER_BANNER

File: libRALFit/src/ral_nlls_internal.f90 [hunk 11]
```diff
@@ -693,162 +774,142 @@ recursive subroutine nlls_iterate(n, m, X,                   &
...
        w%resvec (w%iter + 1) = inform%obj
        w%gradvec(w%iter + 1) = inform%norm_g
     end if
-    
-    if (options%print_level >=1) then
-       write(options%out,1010) w%iter, second, w%Delta, rho, inform%obj, &
-            inform%norm_g, inform%scaled_g
-    end if
+
+!   @TAG:ITERBNR
```
Comment (talassio) [inline libRALFit/src/ral_nlls_internal.f90]: tag changed to ITER_BANNER, added comments to describe code segment and output columns

File: libRALFit/src/ral_nlls_internal.f90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: Columns are now described in tag: ITER_BANNER.

[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: Issue #22 was raised and assigned

File: libRALFit/src/ral_nlls_internal.f90 [hunk 12]
```diff
@@ -3113,63 +3450,73 @@ subroutine get_svd_J(n,m,J,s1,sn,options,inform,status,w)
        jobu  = 'N' ! calculate no left singular vectors
        jobvt = 'N' ! calculate no right singular vectors
        lwork = size(w%work)
-! TODO AndrewS: Lanczos.f90 causes Warning: Floating divide by zero occurred
+! TODO AndrewS: Lanczos.f90 causes Warning: Floating divide by 0.0_wp occurred
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/src/ral_nlls_internal.f90]: Comments removed and raised issue #23

File: libRALFit/CMakeLists.txt [hunk 13]
```diff
@@ -49,20 +52,28 @@ if(CMAKE_BUILD_TYPE_LOWER MATCHES debug)
   set(TestCoverage On)
 elseif(CMAKE_BUILD_TYPE_LOWER MATCHES profile)
       message("WARNING: Debug version - not to be used in production code.")	
-  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} -pg -g -O3")
```
[the remark this answers is not in the document]
Comment (talassio) [reply libRALFit/CMakeLists.txt]: `update_cmake` branch was merged

**Comment by tyronerees** [comment]
The changes to the codes look good -- thanks, Andrew!  

These changes have dropped the test coverage percentage significantly.  In particular, the tests don't hit the new printing routines you've written.  Could you amend `test/nlls_test.f90` to ensure that all the various branches of the printing code is hit when this is run?  I write `options%out` to `nlls_test.out` and `options%error` to `nlls_test_error.out` there, and I change the print levels at various points throughout the file to ensure everything is hit at least once.

As discussed, we still need the docs to be updated before this can be merged in to master, as they should accurately describe the code.  I see the you have added issue #22 to do this, which is great.  Sphinx (https://docs.readthedocs.io/en/latest/intro/getting-started-with-sphinx.html) is pretty easy to change, but let me know if you have any trouble.

**Comment by talassio** [comment]
> The changes to the codes look good -- thanks, Andrew!
> 
> These changes have dropped the test coverage percentage significantly. In particular, the tests don't hit the new printing routines you've written. Could you amend `test/nlls_test.f90` to ensure that all the various branches of the printing code is hit when this is run? I write `options%out` to `nlls_test.out` and `options%error` to `nlls_test_error.out` there, and I change the print levels at various points throughout the file to ensure everything is hit at least once.
> 
> As discussed, we still need the docs to be updated before this can be merged in to master, as they should accurately describe the code. I see the you have added issue #22 to do this, which is great. Sphinx (https://docs.readthedocs.io/en/latest/intro/getting-started-with-sphinx.html) is pretty easy to change, but let me know if you have any trouble.

Updated nlls_test and now it covers all the printing module, coverage still does not hit target due to allocation checks in nlls_workspaces.

**Comment by tyronerees** [comment]
A few more relatively minor points I've found:

**Docs**
In the Fortran docs:
- [ ] `print_options` is logical, but listed as integer
- [ ] `print_header` is integer, but listed as logical

In all docs:
- [ ] the description of `print_header` should read "prints the column header every `print_header` iterations when `print_level > 1`" (not have the value 30 hard-coded in)

**Python interface** 
The python interface breaks with this update.  The file `src/ral_nlls_pyiface.c` should be updated to match the current option list.

The python interface is not currently included in the test -- I've added Issue #30 to address this.  I've also discovered that the python instructions don't work on all systems (as it needs the numpy headers to be in the path) -- I've addressed this with issue #29, which is now fixed and merged into master.

**Tests** 
Can the if statements before the calls to remove workspaces be inlined to increase test coverage?  In particular, at:
- [call to remove_workspaces](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L809)
- [call to remove_workspace_get_svd](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L903)
- [call to remove_workspace_dogleg](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1184)
- [call to remove_workspace_AINT_tr](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1299)
- [call to remove_workspace_min_eig_symm](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1411)
- [call to remove_workspace_max_eig](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1488)
- [call to remove_workspace_galahad](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1559)
- [deallocating w%ew](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1621)
- [call to remove_workspace_all_eig_symm](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1658)
- [status check on setup_workspace_min_eig_symm](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1689)
- [call to remove_workspace_more_sorensen](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1696)
- [call to remove_workspace_generate_scaling](https://github.com/talassio/RALFit/blob/2fdbe9b5570941578046ec092a1935384fc856ca/libRALFit/src/ral_nlls_workspaces.f90#L1745)

The error return `NLLS_ERROR_WRONG_INNER_METHOD` is not checked in the tests

**Pull request merged**